packages/gatsby/src/internal-plugins/webpack-theme-component-shadowing/index.js

```diff
@@ -15,147 +23,165 @@ const pathWithoutExtension = fullPath => {
-          request.context.issuer &&
-          /**
-           * An issuer is the file making the require request. It can
-           * be in a user's site or a theme. If the issuer is requesting
-           * a path in the shadow chain that it participates in, then we
-           * will let the request through as normal. Otherwise, we
-           * engage the shadowing algorithm.
-           */
-          this.requestPathIsIssuerShadowPath({
-            requestPath: request.path,
-            issuerPath: request.context.issuer,
-            userSiteDir: this.projectRoot,
+    resolver
+      .getHook(`before-resolved`)
+      .tapAsync(
+        `GatsbyThemeComponentShadowingResolverPlugin`,
+        (request, stack, callback) => {
+          const [theme, component] = this.getThemeAndComponent(request.path)
+          if (!theme) {
+            return callback()
+          }
+
+          if (
+            /**
+             * if someone adds
+             * ```
+             * modules: [path.resolve(__dirname, /src'), 'node_modules'],
+             * ```
+             * to the webpack config, `issuer` is `null`, so we skip this check.
+             * note that it's probably a bad idea in general to set `modules`
+             * like this in a theme, but we also shouldn't artificially break
+             * people that do.
+             */
+            request.context.issuer &&
+            /**
+             * An issuer is the file making the require request. It can
+             * be in a user's site or a theme. If the issuer is requesting
+             * a path in the shadow chain that it participates in, then we
+             * will let the request through as normal. Otherwise, we
+             * engage the shadowing algorithm.
+             */
+            this.requestPathIsIssuerShadowPath({
+              component,
+              theme,
+              issuerPath: request.context.issuer,
+              userSiteDir: this.projectRoot,
+            })
+          ) {
+            return callback()
+          }
+
+          // This is the shadowing algorithm.
+          const builtComponentPath = this.resolveComponentPath({
+            theme,
+            component,
           })
-        ) {
-          return resolver.doResolve(
-            resolver.hooks.describedRelative,
-            request,
-            null,
-            {},
-            callback
-          )
-        }
 
-        // This is the shadowing algorithm.
-        const builtComponentPath = this.resolveComponentPath({
-          matchingTheme: theme.themeName,
-          themes: this.themes,
-          component,
-        })
-
-        return resolver.doResolve(
-          resolver.hooks.describedRelative,
-          { ...request, path: builtComponentPath || request.path },
-          null,
-          {},
-          callback
-        )
-      }
-    )
+          if (builtComponentPath) {
+            return resolver.doResolve(
+              resolver.hooks.describedRelative,
```

Comment:
is this correct to use same hook as argument despite tapping into different one?

Reply:
Well... This is indeed _legal_ and this is done extensively in Webpack's default resolve pipeline.

Now, if your question is either this is the most appropriate hook to jump back to, that I have to admit I am not certain of it. Given that we are providing a fully resolved path to the shadowing file, I don't think there is anything left that ought to be done by the default resolve pipeline.

But it is hard to be sure here, because Webpack provides no official documentation of these hooks and which work is performed at which step, aside from the source code itself (mostly https://github.com/webpack/enhanced-resolve/blob/main/lib/ResolverFactory.js and implementations of the various plugins named there), which may change in the future.

Still, `describedRelative` appears to be a reasonably safe and stable point to reintroduce altered paths in the middle of the resolve pipeline, making it more future-proof. That is why I kept that one specifically.

Reply:
> But it is hard to be sure here, because Webpack provides no official documentation of these hooks and which work is performed at which step

Yeah, this part is what I find most difficult when working with webpack plugins/loaders etc. Unless you are very familiar with how webpack does things internally (by doing deep dives in sourcecode and trying to "reverse engineer" what and how is meant to work). 

I just wasn't sure here if what we had previously (using `describedRelative` in `relative` hook) was intentionally "the same" or using `describedRelative` just make sense in more scenarios than that. I'm completely fine with keeping it as-is if it make sense.